File: clients/python/clausters/seq/pattern.py
"""Patterns (port of ``sc3/seq/pattern.py`` + ``patterns/``).

A `Pattern` is a reusable, lazy description of a value sequence -- the
definition of a generator: iterating it yields the values (a fresh stream each
time), and it does not play. Value patterns (``Pseq``, ``Pwhite``, ...) feed
`Pbind`, which combines per-key value patterns into a stream of `Event`
objects. An `EventPattern` -- `Pbind`, and a list pattern whose every element
is an event pattern -- is what plays, on a clock, with `EventPattern.play` (see
`EventStreamPlayer`).

Patterns are plain Python generators under the hood, so nesting and composition
are natural; a sub-pattern used as a value is embedded (iterated) in place.
"""

import math

from ..base import rand
from .event import Event

INF = math.inf


def _embed(value):
    """Yields a value, or iterates it if it is itself a pattern."""
    if isinstance(value, Pattern):
        yield from iter(value)
    else:
        yield value


def as_pattern(value):
    return value if isinstance(value, Pattern) else Pconst(value)


def _made(cls, base, event_cls, events):
    """An instance of ``event_cls`` when ``base`` itself is being built over
    event patterns only, else of ``cls``: how a list pattern resolves whether
    it is an `EventPattern` when it is made."""
    return object.__new__(event_cls if cls is base and events else cls)


def _all_events(items) -> bool:
    return bool(items) and all(isinstance(item, EventPattern) for item in items)


class Pattern:
    """The definition of a generator: iterating it yields its values. It has no
    `play` -- what plays is an `EventPattern`. `clausters.render` of a value
    pattern generates its values."""

    def __iter__(self):
        raise NotImplementedError(f"{type(self).__name__}.__iter__")

    def stream(self):
        """A `Stream` over this pattern."""
        from ..base.stream import FunctionStream

        it = iter(self)
        return FunctionStream(lambda _=None: next(it))


class EventPattern(Pattern):
    """A pattern whose values are events, and so a pattern that plays.

    `Pbind` is one, and so is a list pattern (`Pseq`, `Prand`, `Pn`) whose
    every element is one: it resolves its class when it is built.
    """

    def play(self, clock=None, server=None, quant=None):
        """Play this event pattern on ``clock``, sending to ``server``.

        Both are optional and resolve against the ambient context (the running
        session, else the default session): ``server=None`` takes the booted
        default server, and ``clock=None`` takes the running routine's clock or,
        outside one, the default session's clock -- created and started on first
        use. So ``Pbind(...).play()`` sounds with only a ``Server().boot()`` and no
        `Session`."""
        from ..base.main import main
        from .eventstream import EventStreamPlayer

        server = main.resolve_server(server)
        if clock is None:
            clock = main.resolve_clock() or main.get_default_clock()
        return EventStreamPlayer(self, server).play(clock, quant)


# ---- value patterns ----

class Pconst(Pattern):
    """A constant value, ``length`` times (infinite by default)."""

    def __init__(self, value, length=INF):
        self.value = value
        self.length = length

    def __iter__(self):
        i = 0
        while self.length is INF or i < self.length:
            yield self.value
            i += 1


class Pseq(Pattern):
    """The items in order, ``repeats`` times (sub-patterns are embedded). Over
    event patterns only, it is an `EventPattern`."""

    def __new__(cls, items=(), repeats=1):
        items = list(items)
        self = _made(cls, Pseq, _EventPseq, _all_events(items))
        self.items = items
        return self

    def __init__(self, items=(), repeats=1):
        self.repeats = repeats

    def __iter__(self):
        i = 0
        while self.repeats is INF or i < self.repeats:
            for item in self.items:
                yield from _embed(item)
            i += 1


class Pser(Pattern):
    """The items in order, yielding exactly ``length`` values (cycling)."""

    def __init__(self, items, length):
        self.items = list(items)
        self.length = length

    def __iter__(self):
        for i in range(int(self.length)):
            yield self.items[i % len(self.items)]


class Prand(Pattern):
    """Random items, ``length`` values, drawn from the **random context** (the
    running routine's generator, or the root outside one -- see
    `clausters.base.rand`): ``main.seed(n)`` reproduces the choices along with
    everything else in the script. There is no per-pattern seed -- independent
    seeds would break whole-script consistency. Over event patterns only, it is
    an `EventPattern`."""

    def __new__(cls, items=(), length=INF):
        items = list(items)
        self = _made(cls, Prand, _EventPrand, _all_events(items))
        self.items = items
        return self

    def __init__(self, items=(), length=INF):
        self.length = length

    def __iter__(self):
        i = 0
        while self.length is INF or i < self.length:
            yield from _embed(rand.choice(self.items))
            i += 1


class Pwhite(Pattern):
    """Uniform random numbers in ``[lo, hi)``, ``length`` values, drawn from
    the **random context** (the running routine's generator, or the root
    outside one -- see `clausters.base.rand`): ``main.seed(n)`` reproduces the
    sequence along with everything else in the script. There is no per-pattern
    seed -- independent seeds would break whole-script consistency."""

    def __init__(self, lo=0.0, hi=1.0, length=INF):
        self.lo, self.hi, self.length = lo, hi, length

    def __iter__(self):
        i = 0
        while self.length is INF or i < self.length:
            yield rand.uniform(self.lo, self.hi)
            i += 1


class Pseries(Pattern):
    """Arithmetic series ``start, start+step, ...`` (``length`` values)."""

    def __init__(self, start=0.0, step=1.0, length=INF):
        self.start, self.step, self.length = start, step, length

    def __iter__(self):
        value = self.start
        i = 0
        while self.length is INF or i < self.length:
            yield value
            value += self.step
            i += 1


class Pgeom(Pattern):
    """Geometric series ``start, start*grow, ...`` (``length`` values)."""

    def __init__(self, start=1.0, grow=2.0, length=INF):
        self.start, self.grow, self.length = start, grow, length

    def __iter__(self):
        value = self.start
        i = 0
        while self.length is INF or i < self.length:
            yield value
            value *= self.grow
            i += 1


class Pfunc(Pattern):
    """Calls ``func()`` for each value (``length`` values)."""

    def __init__(self, func, length=INF):
        self.func = func
        self.length = length

    def __iter__(self):
        i = 0
        while self.length is INF or i < self.length:
            yield self.func()
            i += 1


class Pn(Pattern):
    """Repeats ``pattern`` ``n`` times. Over an event pattern, it is an
    `EventPattern`."""

    def __new__(cls, pattern=None, n=INF):
        return _made(cls, Pn, _EventPn, isinstance(pattern, EventPattern))

    def __init__(self, pattern=None, n=INF):
        self.pattern = pattern
        self.n = n

    def __iter__(self):
        i = 0
        while self.n is INF or i < self.n:
            yield from _embed(self.pattern)
            i += 1


# ---- event pattern ----

class Pbind(EventPattern):
    """Binds keys to value patterns; yields an `Event` per step, stopping
    when any key's stream stops. Constant values are held; sub-patterns advance
    one value per event.

    Constructs exactly like a ``dict`` -- keyword arguments
    (``Pbind(freq=Pseq([440, 880]), dur=0.5)``), a mapping
    (``Pbind({"freq": Pseq([440, 880]), "dur": 0.5})``), or both merged, with
    keywords winning. The dict form also admits keys that are not valid Python
    identifiers."""

    def __init__(self, *args, **patterns):
        self.patterns = dict(*args, **patterns)

    def __iter__(self):
        streams = {key: iter(as_pattern(value)) for key, value in self.patterns.items()}
        while True:
            event = {}
            for key, stream in streams.items():
                try:
                    event[key] = next(stream)
                except StopIteration:
                    return
            yield Event(event)


# ---- list patterns over events ----
#
# What a list pattern is made as when every element is an event pattern: the
# same pattern, and playable. `Pseq([Pbind(...), Pbind(...)])` is one.

class _EventPseq(Pseq, EventPattern):
    pass


class _EventPrand(Prand, EventPattern):
    pass


class _EventPn(Pn, EventPattern):
    pass

File: clients/web/src/seq/pattern.ts
// Patterns (mirrors `clausters/seq/pattern.py`).
//
// A `Pattern` is a reusable, lazy description of a value sequence -- the
// definition of a generator: iterating it yields the values (a fresh walk each
// time), and it does not play. Value patterns (`Pseq`, `Pwhite`, ...) feed
// `Pbind`, which combines per-key value patterns into a stream of `Event`s. An
// `EventPattern` -- `Pbind`, and a list pattern whose every element is an event
// pattern -- is what plays, on a clock, with `EventPattern.play` (see
// `EventStreamPlayer`).
//
// Patterns are plain generators underneath, so nesting and composition are
// natural; a sub-pattern used as a value is embedded (iterated) in place.

import { FunctionStream, StopStream } from "../base/stream.ts";
import type { Stream } from "../base/stream.ts";
import * as rand from "../base/rand.ts";
import { main } from "../base/main.ts";
import { Event } from "./event.ts";
import type { EventDestination } from "./event.ts";
import { EventStreamPlayer } from "./eventstream.ts";
import type { TempoClock } from "../base/clock.ts";

/** An endless pattern's length. */
export const INF = Infinity;

/** Yields a value, or iterates it if it is itself a pattern. */
function* embed<T>(value: T | Pattern<T>): Generator<T, void, undefined> {
    if (value instanceof Pattern) yield* value;
    else yield value;
}

/** A bare value used where a pattern is expected becomes an endless constant. */
export const asPattern = <T>(value: T | Pattern<T>): Pattern<T> =>
    value instanceof Pattern ? value : new Pconst(value);

/**
 * The definition of a generator: walking it yields its values. It has no
 * `play` -- what plays is an {@link EventPattern}. `render` of a value pattern
 * generates its values.
 */
export abstract class Pattern<T = unknown> {
    abstract [Symbol.iterator](): Generator<T, void, undefined>;

    /** A `Stream` over this pattern -- the form the clock can resume. */
    stream(): Stream {
        const it = this[Symbol.iterator]();
        return new FunctionStream(() => {
            const step = it.next();
            if (step.done) throw new StopStream();
            return step.value;
        });
    }
}

/** What an event pattern carries, so a list pattern built over events is one. */
const EVENTS = Symbol("clausters.eventPattern");

/** Plays `pattern`'s events on `clock`, sending to `destination`. */
function playEvents(
    pattern: Pattern<unknown>,
    destination: EventDestination | undefined,
    { clock, quant }: { clock?: TempoClock; quant?: number },
): EventStreamPlayer {
    const target = destination ?? (main.resolveServer() as unknown as EventDestination);
    return new EventStreamPlayer(pattern, target).play(clock, quant);
}

/**
 * A pattern whose values are events, and so a pattern that plays.
 *
 * `Pbind` is one, and so is a list pattern (`Pseq`, `Prand`, `Pn`) whose every
 * element is one: it resolves what it is when it is built, and
 * `instanceof EventPattern` answers for both.
 */
export abstract class EventPattern<T = Event> extends Pattern<T> {
    readonly [EVENTS] = true;

    static override [Symbol.hasInstance](value: unknown): boolean {
        if (this !== EventPattern) return Function.prototype[Symbol.hasInstance].call(this, value);
        return value instanceof Pattern && (value as { [EVENTS]?: boolean })[EVENTS] === true;
    }

    /**
     * Plays this event pattern on `clock`, sending to `destination`.
     *
     * Both are optional and resolve against the ambient context (the running
     * session, else the active one, else the default session): an omitted
     * `destination` takes that environment's server, and an omitted `clock`
     * the running routine's or, outside one, the default session's -- created
     * and started on first use. So `new Pbind(...).play()` sounds with a
     * `Session` opened somewhere and nothing else wired.
     */
    play(
        destination?: EventDestination,
        options: { clock?: TempoClock; quant?: number } = {},
    ): EventStreamPlayer {
        return playEvents(this as Pattern<unknown>, destination, options);
    }
}

/** Whether every element of a list pattern is an event pattern. */
const allEvents = (items: readonly unknown[]): boolean =>
    items.length > 0 && items.every((item) => item instanceof EventPattern);

// ---- value patterns ----

/** A constant value, `length` times (endless by default). */
export class Pconst<T> extends Pattern<T> {
    readonly value: T;
    readonly length: number;

    constructor(value: T, length: number = INF) {
        super();
        this.value = value;
        this.length = length;
    }

    *[Symbol.iterator](): Generator<T, void, undefined> {
        for (let i = 0; i < this.length; i++) yield this.value;
    }
}

/**
 * The items in order, `repeats` times (sub-patterns are embedded). Over event
 * patterns only, it is an {@link EventPattern}.
 */
export class Pseq<T> extends Pattern<T> {
    readonly items: readonly (T | Pattern<T>)[];
    readonly repeats: number;

    constructor(items: readonly (T | Pattern<T>)[], repeats: number = 1) {
        super();
        this.items = [...items];
        this.repeats = repeats;
        if (new.target === Pseq && allEvents(this.items)) {
            return new EventPseq(this.items, repeats);
        }
    }

    *[Symbol.iterator](): Generator<T, void, undefined> {
        for (let i = 0; i < this.repeats; i++) {
            for (const item of this.items) yield* embed(item);
        }
    }
}

/** The items in order, yielding exactly `length` values (cycling). */
export class Pser<T> extends Pattern<T> {
    readonly items: readonly T[];
    readonly length: number;

    constructor(items: readonly T[], length: number) {
        super();
        this.items = [...items];
        this.length = length;
    }

    *[Symbol.iterator](): Generator<T, void, undefined> {
        for (let i = 0; i < this.length; i++) yield this.items[i % this.items.length]!;
    }
}

/**
 * Random items, `length` values, drawn from the **random context** (the
 * running routine's stream, or the root outside one -- see `base/rand.ts`):
 * `seed(n)` reproduces the choices along with everything else in the script.
 * There is no per-pattern seed -- independent seeds would break whole-script
 * consistency. Over event patterns only, it is an {@link EventPattern}.
 */
export class Prand<T> extends Pattern<T> {
    readonly items: readonly (T | Pattern<T>)[];
    readonly length: number;

    constructor(items: readonly (T | Pattern<T>)[], length: number = INF) {
        super();
        this.items = [...items];
        this.length = length;
        if (new.target === Prand && allEvents(this.items)) {
            return new EventPrand(this.items, length);
        }
    }

    *[Symbol.iterator](): Generator<T, void, undefined> {
        for (let i = 0; i < this.length; i++) yield* embed(rand.choice(this.items));
    }
}

/**
 * Uniform random numbers in `[lo, hi)`, `length` values, from the random
 * context (see `Prand` on seeding).
 */
export class Pwhite extends Pattern<number> {
    readonly lo: number;
    readonly hi: number;
    readonly length: number;

    constructor(lo = 0, hi = 1, length: number = INF) {
        super();
        this.lo = lo;
        this.hi = hi;
        this.length = length;
    }

    *[Symbol.iterator](): Generator<number, void, undefined> {
        for (let i = 0; i < this.length; i++) yield rand.uniform(this.lo, this.hi);
    }
}

/** Arithmetic series `start, start + step, ...` (`length` values). */
export class Pseries extends Pattern<number> {
    readonly start: number;
    readonly step: number;
    readonly length: number;

    constructor(start = 0, step = 1, length: number = INF) {
        super();
        this.start = start;
        this.step = step;
        this.length = length;
    }

    *[Symbol.iterator](): Generator<number, void, undefined> {
        let value = this.start;
        for (let i = 0; i < this.length; i++) {
            yield value;
            value += this.step;
        }
    }
}

/** Geometric series `start, start * grow, ...` (`length` values). */
export class Pgeom extends Pattern<number> {
    readonly start: number;
    readonly grow: number;
    readonly length: number;

    constructor(start = 1, grow = 2, length: number = INF) {
        super();
        this.start = start;
        this.grow = grow;
        this.length = length;
    }

    *[Symbol.iterator](): Generator<number, void, undefined> {
        let value = this.start;
        for (let i = 0; i < this.length; i++) {
            yield value;
            value *= this.grow;
        }
    }
}

/** Calls `func()` for each value (`length` values). */
export class Pfunc<T> extends Pattern<T> {
    readonly func: () => T;
    readonly length: number;

    constructor(func: () => T, length: number = INF) {
        super();
        this.func = func;
        this.length = length;
    }

    *[Symbol.iterator](): Generator<T, void, undefined> {
        for (let i = 0; i < this.length; i++) yield this.func();
    }
}

/** Repeats `pattern` `n` times. Over an event pattern, it is an {@link EventPattern}. */
export class Pn<T> extends Pattern<T> {
    readonly pattern: T | Pattern<T>;
    readonly n: number;

    constructor(pattern: T | Pattern<T>, n: number = INF) {
        super();
        this.pattern = pattern;
        this.n = n;
        if (new.target === Pn && pattern instanceof EventPattern) {
            return new EventPn(pattern, n);
        }
    }

    *[Symbol.iterator](): Generator<T, void, undefined> {
        for (let i = 0; i < this.n; i++) yield* embed(this.pattern);
    }
}

// ---- event pattern ----

/**
 * The keys a `Pbind` binds: each one a pattern, or a constant held for every
 * event.
 */
export type Bindings = Record<string, unknown>;

/**
 * Binds keys to value patterns; yields an `Event` per step, stopping when any
 * key's walk stops. Constants are held; sub-patterns advance one value per
 * event.
 */
export class Pbind extends EventPattern<Event> {
    readonly patterns: Bindings;

    constructor(patterns: Bindings) {
        super();
        this.patterns = { ...patterns };
    }

    *[Symbol.iterator](): Generator<Event, void, undefined> {
        const walks = Object.entries(this.patterns).map(
            ([key, value]) =>
                [key, asPattern(value)[Symbol.iterator]()] as const,
        );
        for (;;) {
            const props: Record<string, unknown> = {};
            for (const [key, walk] of walks) {
                const step = walk.next();
                if (step.done) return;
                props[key] = step.value;
            }
            yield new Event(props);
        }
    }
}

// ---- list patterns over events ----
//
// What a list pattern is made as when every element is an event pattern: the
// same pattern, and playable. `new Pseq([new Pbind(...), new Pbind(...)])` is one.

class EventPseq<T> extends Pseq<T> {
    readonly [EVENTS] = true;

    play(destination?: EventDestination, options: { clock?: TempoClock; quant?: number } = {}): EventStreamPlayer {
        return playEvents(this as Pattern<unknown>, destination, options);
    }
}

class EventPrand<T> extends Prand<T> {
    readonly [EVENTS] = true;

    play(destination?: EventDestination, options: { clock?: TempoClock; quant?: number } = {}): EventStreamPlayer {
        return playEvents(this as Pattern<unknown>, destination, options);
    }
}

class EventPn<T> extends Pn<T> {
    readonly [EVENTS] = true;

    play(destination?: EventDestination, options: { clock?: TempoClock; quant?: number } = {}): EventStreamPlayer {
        return playEvents(this as Pattern<unknown>, destination, options);
    }
}
